cmd/simulator/load/worker.go

```diff
@@ -10,102 +10,75 @@ import (
-func NewWorker(client ethclient.Client, address common.Address, txs []*types.Transaction) *Worker {
-	return &Worker{
-		client:  client,
-		address: address,
-		txs:     txs,
-	}
-}
+	acceptedNonce uint64
+	address       common.Address
 
-func (w *Worker) ExecuteTxsFromAddress(ctx context.Context) error {
-	log.Info("Executing txs", "numTxs", len(w.txs))
-	for i, tx := range w.txs {
-		start := time.Now()
-		err := w.client.SendTransaction(ctx, tx)
-		if err != nil {
-			return fmt.Errorf("failed to issue tx %d: %w", i, err)
-		}
-		log.Info("execute tx", "tx", tx.Hash(), "nonce", tx.Nonce(), "duration", time.Since(start))
-	}
-	return nil
+	sub      interfaces.Subscription
+	newHeads chan *types.Header
 }
 
-// AwaitTxs awaits for the nonce of the last transaction issued by the worker to be confirmed or
-// rejected by the network.
-//
-// Assumes that a non-zero number of transactions were already generated and that they were issued
-// by this worker.
-func (w *Worker) AwaitTxs(ctx context.Context) error {
-	nonce := w.txs[len(w.txs)-1].Nonce()
-
+// what do we want
+// ongoing goroutine that updates the nonce and sends/drops an event if there is any listener
+func NewSingleAddressTxWorker(ctx context.Context, client ethclient.Client, address common.Address) *singleAddressTxWorker {
 	newHeads := make(chan *types.Header)
-	defer close(newHeads)
+	tw := &singleAddressTxWorker{
+		client:   client,
+		address:  address,
+		newHeads: newHeads,
+	}
 
-	sub, err := w.client.SubscribeNewHead(ctx, newHeads)
+	sub, err := client.SubscribeNewHead(ctx, newHeads)
 	if err != nil {
 		log.Debug("failed to subscribe new heads, falling back to polling", "err", err)
 	} else {
-		defer sub.Unsubscribe()
+		tw.sub = sub
 	}
 
+	return tw
+}
+
+func (tw *singleAddressTxWorker) IssueTx(ctx context.Context, tx *types.Transaction) error {
+	return tw.client.SendTransaction(ctx, tx)
+}
+
+func (tw *singleAddressTxWorker) ConfirmTx(ctx context.Context, tx *types.Transaction) error {
+	txNonce := tx.Nonce()
+
 	for {
+		// If the is less than what has already been accepted, the transaction is confirmed
+		if txNonce < tw.acceptedNonce {
+			return nil
+		}
+
 		select {
-		case <-newHeads:
+		case <-tw.newHeads:
 		case <-time.After(time.Second):
 		case <-ctx.Done():
-			return fmt.Errorf("failed to await nonce: %w", ctx.Err())
+			return fmt.Errorf("failed to await tx %s nonce %d: %w", tx.Hash(), txNonce, ctx.Err())
 		}
 
-		currentNonce, err := w.client.NonceAt(ctx, w.address, nil)
+		// Update the worker's accepted nonce, so we can check on the next iteration
+		// if the transaction has been accepted.
+		acceptedNonce, err := tw.client.NonceAt(ctx, tw.address, nil)
 		if err != nil {
-			log.Warn("failed to get nonce", "err", err)
-		}
-		if currentNonce >= nonce {
-			return nil
-		} else {
-			log.Info("fetched nonce", "awaiting", nonce, "currentNonce", currentNonce)
+			return fmt.Errorf("failed to await tx %s nonce %d: %w", tx.Hash(), txNonce, err)
 		}
```

Comment:
can we still keep the awaiting line. It's helpful to the user.

Reply:
I think it's excessive to do so for every nonce at the info level. Could we make it debug?